network/p2p/libp2pNode.go

```diff
@@ -290,59 +312,53 @@ func (n *Node) Stop() (chan struct{}, error) {
 }
 
 // AddPeer adds a peer to this node by adding it to this node's peerstore and connecting to it
-func (n *Node) AddPeer(ctx context.Context, identity flow.Identity) error {
-	pInfo, err := PeerAddressInfo(identity)
-	if err != nil {
-		return fmt.Errorf("failed to add peer %s: %w", identity.String(), err)
-	}
-
-	err = n.host.Connect(ctx, pInfo)
+func (n *Node) AddPeer(ctx context.Context, peerInfo peer.AddrInfo) error {
+	err := n.host.Connect(ctx, peerInfo)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
-// RemovePeer closes the connection with the identity.
-func (n *Node) RemovePeer(ctx context.Context, identity flow.Identity) error {
-	pInfo, err := PeerAddressInfo(identity)
-	if err != nil {
-		return fmt.Errorf("failed to remove peer %x: %w", identity, err)
-	}
-
-	err = n.host.Network().ClosePeer(pInfo.ID)
+// RemovePeer closes the connection with the peer.
+func (n *Node) RemovePeer(ctx context.Context, peerID peer.ID) error {
+	err := n.host.Network().ClosePeer(peerID)
 	if err != nil {
-		return fmt.Errorf("failed to remove peer %s: %w", identity, err)
+		return fmt.Errorf("failed to remove peer %s: %w", peerID, err)
 	}
 	return nil
 }
 
-// CreateStream returns an existing stream connected to identity, if it exists or adds one to identity as a peer and creates a new stream with it.
-func (n *Node) CreateStream(ctx context.Context, identity flow.Identity) (libp2pnet.Stream, error) {
+// CreateStream returns an existing stream connected to the peer if it exists, or creates a new stream with it.
+func (n *Node) CreateStream(ctx context.Context, peerID peer.ID) (libp2pnet.Stream, error) {
+	if len(n.host.Peerstore().Addrs(peerID)) == 0 {
+		n.logger.Info().Str("peerID", peerID.Pretty()).Msg("address not found in peerstore")
+		if n.dht != nil {
+			n.logger.Info().Str("peerID", peerID.Pretty()).Msg("searching for peer in dht")
+			timedCtx, cancel := context.WithTimeout(ctx, findPeerQueryTimeout)
+			// try to find the peer using the dht
+			_, err := n.dht.FindPeer(timedCtx, peerID)
+			cancel()
+			if err != nil {
+				return nil, fmt.Errorf("could not find address for peer %v: %w", peerID, err)
+			}
+			n.logger.Info().Str("peerID", peerID.Pretty()).Msg("address found")
+		}
+		return nil, fmt.Errorf("no valid addresses exist for peer %v", peerID)
+	}
 	// Open libp2p Stream with the remote peer (will use an existing TCP connection underneath if it exists)
-	stream, err := n.tryCreateNewStream(ctx, identity, maxConnectAttempt)
+	stream, err := n.tryCreateNewStream(ctx, peerID, maxConnectAttempt)
 	if err != nil {
-		return nil, flownet.NewPeerUnreachableError(fmt.Errorf("could not create stream (node_id: %s, address: %s): %w", identity.NodeID.String(),
-			identity.Address, err))
+		return nil, flownet.NewPeerUnreachableError(fmt.Errorf("could not create stream (peer_id: %s): %w", peerID, err))
```

Comment:
since PeerID are fairly different than Flow identity, these errors make it difficult to realize which node was not reachable. Currently, [these errors](https://dapperlabs.grafana.net/explore?orgId=1&left=%5B%22now-15m%22,%22now%22,%22grafanacloud-dapperlabs-logs%22,%7B%22expr%22:%22%7Bnetwork%3D%5C%22mainnet12%5C%22%7D%20%7C%3D%20%5C%22could%20not%20create%20stream%5C%22%22,%22requestId%22:%22Q-07a90617-e9e5-44f8-b407-732a1eee7312-0A%22%7D%5D) are reported as e.g.,

> "failed to send message to b4f850b43df502e7fd292c34d9a21c2e8db350cd8a1214a4a29734bc6f3c9e92: failed to create stream for b4f850b43df502e7fd292c34d9a21c2e8db350cd8a1214a4a29734bc6f3c9e92 :could not create stream (node_id: b4f850b43df502e7fd292c34d9a21c2e8db350cd8a1214a4a29734bc6f3c9e92, address: 195.201.31.78:3569): remote node is running on a different spork: protocol not supported, protocol attempted: /flow/push/99b758b89f7a5d30478125abac8da596df7ff58fc52e853b4433d3174a911731"

These happen a lot and the address really helps in easy debugging. Is there a way we could print  the node id and/or the address here?

Reply:
So, If we want to print the node ID, then it should be done at the level up, ie whoever is calling `CreateStream`. I can make sure that when the error is printed at that level, the node ID is included.

Reply:
@vishalchangrani actually it is already here: https://github.com/onflow/flow-go/blob/smnzhu%2Fidentity-provider/network/p2p/middleware.go#L308

And as you can notice in the error message example you pasted, the node ID is printed three times already in the message, while the peerID is not present anywhere.